### PR TITLE
Allow platform frameworks to snapshot FlutterView contents.

### PR DESCRIPTION
**Pull request by chinmaygarde**
Fixes https://github.com/flutter/flutter/issues/6711

### CODE REVIEW

**Comment by abarth** [comment]
LGTM

File: shell/platform/darwin/ios/framework/Source/FlutterView.mm [hunk 1]
```diff
@@ -30,4 +36,87 @@ - (BOOL)enableInputClicksWhenVisible {
   return YES;
 }
 
+void SnapshotRasterizer(ftl::WeakPtr<shell::Rasterizer> rasterizer,
+                        CGContextRef context,
+                        bool is_opaque) {
+  if (!rasterizer) {
```
Comment (collinjackson) [inline shell/platform/darwin/ios/framework/Source/FlutterView.mm]: I think the style for this part of the code base is to not use braces for single-line if statements.

File: shell/platform/darwin/ios/framework/Source/FlutterView.mm [hunk 2]
```diff
@@ -30,4 +36,87 @@ - (BOOL)enableInputClicksWhenVisible {
...
+    return;
+  }
+
+  // Access the layer tree and assess the description of the backing store to
+  // create for this snapshot.
+  flow::LayerTree* layer_tree = rasterizer->GetLastLayerTree();
+  if (layer_tree == nullptr) {
```
Comment (collinjackson) [inline shell/platform/darwin/ios/framework/Source/FlutterView.mm]: ditto

File: shell/platform/darwin/ios/framework/Source/FlutterView.mm [hunk 3]
```diff
@@ -30,4 +36,87 @@ - (BOOL)enableInputClicksWhenVisible {
...
+    return;
+  }
+  auto size = layer_tree->frame_size();
+  if (size.isEmpty()) {
```
Comment (collinjackson) [inline shell/platform/darwin/ios/framework/Source/FlutterView.mm]: ditto

File: shell/platform/darwin/ios/framework/Source/FlutterView.mm [hunk 4]
```diff
@@ -30,4 +36,87 @@ - (BOOL)enableInputClicksWhenVisible {
...
+  }
+  auto info =
+      SkImageInfo::MakeN32(size.width(), size.height(),
+                           is_opaque ? SkAlphaType::kOpaque_SkAlphaType
+                                     : SkAlphaType::kPremul_SkAlphaType);
+
+  // Create the backing store and prepare for use.
+  SkBitmap bitmap;
+  bitmap.setInfo(info);
+  if (!bitmap.tryAllocPixels()) {
```
Comment (collinjackson) [inline shell/platform/darwin/ios/framework/Source/FlutterView.mm]: ditto

File: shell/platform/darwin/ios/framework/Source/FlutterView.mm [hunk 5]
```diff
@@ -30,4 +36,87 @@ - (BOOL)enableInputClicksWhenVisible {
...
+  shell::Shell::Shared().GetRasterizers(&registered_rasterizers);
+  for (auto& rasterizer : registered_rasterizers) {
+    SnapshotRasterizer(rasterizer, context, is_opaque);
+  }
+}
+
+void SnapshotContentsSync(CGContextRef context, UIView* view) {
+  auto gpu_thread = blink::Threads::Gpu();
+
+  if (!gpu_thread) {
```
Comment (collinjackson) [inline shell/platform/darwin/ios/framework/Source/FlutterView.mm]: ditto

File: shell/platform/darwin/ios/framework/Source/FlutterView.mm (hunk 1 above)
Comment (collinjackson) [inline shell/platform/darwin/ios/framework/Source/FlutterView.mm]: P.S. I could be wrong about this, I can see both styles of code in adjacent files, probably it just depends who wrote it. so feel free to ignore me :)

Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterView.mm]: :D From the style guide "curly braces are not required for single-line statements, but they are allowed if you like them;" Doesn't say what to do when you no feelings about it whatsoever. I think i'll just leave it as is.

File: shell/platform/darwin/ios/framework/Source/FlutterView.mm [hunk 6]
```diff
@@ -30,4 +36,87 @@ - (BOOL)enableInputClicksWhenVisible {
...
+  auto frame = compositor_context.AcquireFrame(nullptr, &canvas,
+                                               false /* instrumentation */);
+  layer_tree->Raster(frame, false /* ignore raster cache. */);
+  canvas.flush();
+
+  // Draw the bitmap to the system provided snapshotting context.
+  SkCGDrawBitmap(context, bitmap, 0, 0);
+}
+
+void ShapshotContents(CGContextRef context, bool is_opaque) {
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterView.mm]: Done.

**Comment by chinmaygarde** [comment]
I tested this and `SkCGDrawBitmap` in Skia likes to invert the context to account for CoreGraphics' coordinate system. We (presumably) don't want that because we are assuming callers will that conversion for us. I'll write my own version of `SkCGDrawBitmap`.

**Comment by chinmaygarde** [comment]
On second though, the Skia conversion is correct because it is accounting for the difference between the Skia's coordinate system and CGs.

**Pull request merged**